feat(docs): add OpenCode plugin guide (#3925)
daytonaio/daytona: apps/docs/src/content/docs/en/guides/opencode/opencode-plugin.mdx

---
title: Run OpenCode with the Daytona Plugin
description: Use the Daytona OpenCode plugin to run each OpenCode session in a Daytona sandbox, with automatic git sync to local branches.
---

import { Image } from 'astro:assets'

import opencodePluginGif from '../../../../../assets/docs/images/opencode-plugin.gif'

> This guide demonstrates how to run the [Daytona OpenCode plugin](https://www.npmjs.com/package/@daytonaio/opencode) which seamlessly integrates Daytona sandboxes and OpenCode. When the plugin is active all agent operations occur in secure sandboxes, with one sandbox per OpenCode session. The plugin also has the ability to sync changes between sandboxes and local git branches.

Reviewer: ```suggestion
This guide demonstrates how to run the [Daytona OpenCode plugin](https://www.npmjs.com/package/@daytonaio/opencode) which integrates Daytona sandboxes and OpenCode. When the plugin is active all agent operations occur in secure sandboxes, with one sandbox per OpenCode session. The plugin also has the ability to sync changes between sandboxes and local Git branches.
```

### 1. Workflow Overview

When you run OpenCode with the Daytona plugin, sandboxes will be created automatically inside OpenCode sessions. Operations such as running code, installing dependencies, and starting servers will all occur in the sandbox.

<Image
  src={opencodePluginGif}
  alt="OpenCode with Daytona plugin — sandbox toast and git branches"
  width={600}
  style="max-width: 100%; height: auto; margin: 1rem 0;"
/>

Sandboxes are preserved until you delete the OpenCode session. If a local git repository is detected, the plugin will sync changes between the sandbox and branches with the `opencode/` prefix. 

### 2. Project Setup

#### Add the Plugin

Add the Daytona plugin to your project by creating or editing `opencode.json` in the project directory:

```json
{
  "$schema": "https://opencode.ai/config.json",
  "plugin": ["@daytonaio/opencode"]
}
```

> OpenCode will download the plugin automatically when it starts. To install the plugin globally instead, edit `~/.config/opencode/opencode.json` and add the same `plugin` entry.

Reviewer: ```suggestion
OpenCode downloads the plugin automatically when it starts. To install the plugin globally instead, edit `~/.config/opencode/opencode.json` and add the same `plugin` entry.
```

#### Configure Environment

This plugin requires a [Daytona account](https://www.daytona.io/) and [Daytona API key](https://app.daytona.io/dashboard/keys) to create sandboxes.

Set your API key:

```bash
export DAYTONA_API_KEY="your-api-key"
```

Or create a `.env` file in your project root:

```env
DAYTONA_API_KEY=your-api-key
```

#### Run OpenCode

:::note[Git required]

> Ensure your project is a git repository to enable syncing changes between the sandbox and your machine.

Reviewer: ```suggestion
Ensure your project is a Git repository to enable syncing changes between the sandbox and your machine.
```

:::

> Initialize git if needed, then start OpenCode:

Reviewer: ```suggestion
Initialize Git if needed, then start OpenCode:
```

```bash
git init
opencode
```

You can now use OpenCode as usual. As you work, you will see notifications in OpenCode indicating sandboxes are being created and changes are being synced to local branches.

To confirm the plugin is working, you can type `pwd` in the chat and you should see a path like `/home/daytona/project`.

To view live logs from the plugin for debugging:

```bash
tail -f ~/.local/share/opencode/log/daytona.log
```

#### Version control

In your project directory, use git to list and check out the branches OpenCode creates:

```bash
git branch
git checkout opencode/1
```

By default, new sessions will start from the branch that was checked out when OpenCode was started. After this, synchronization only goes one way: from the sandbox to your local branch. To start working from a different branch, use git tocheck out that branch and start OpenCode again:

```bash
git checkout opencode/1
opencode
```

> You can run as many OpenCode sessions in parallel as you want, and use git afterwards to review and merge changes.

Reviewer: ```suggestion
You can run as many OpenCode sessions in parallel as you want. Use Git to review and merge changes.
```

        'This session is integrated with a Daytona sandbox.',
        `The main project repository is located at: ${repoPath}.`,
        'Bash commands will run in this directory.',
        'Work in this directory. Do NOT try to use the current working directory of the host system.',
        "When executing long-running commands, use the 'background' option to run them asynchronously.",
        'Before showing a preview URL, ensure the server is running in the sandbox on that port.',
      ].join('\n'),
    )
  },
}
```

#### Session cleanup

When you delete an OpenCode session, OpenCode emits a `session.deleted` event. The session cleanup module listens for it, deletes the corresponding sandbox via the session manager, and shows a notification:

```typescript
return {
  event: async ({ event }) => {
    if (event.type === EVENT_TYPE_SESSION_DELETED) {
      const sessionId = (event as EventSessionDeleted).properties.info.id
      await sessionManager.deleteSandbox(sessionId, projectId)
      toast.show({
        title: 'Session deleted',
        message: 'Sandbox deleted successfully.',
        variant: 'success',
      })
    }
  },
}
```

#### File synchronization

> While OpenCode is in use, the plugin uses git to keep session sandboxes and your local git repository in sync. This only occurs if a git repository is detected in the project directory. On plugin start:

Reviewer: ```suggestion
While OpenCode is in use, the plugin uses Git to keep session sandboxes and your local Git repository in sync. This only occurs if a git repository is detected in the project directory. On plugin start:
```
